### ISSUE TITLE
Disconnecting client from server-side too quickly leaves client alive

### ISSUE THREAD

**Issue by joshrickert** [socketio/socket.io#5088]
*Note*: for support questions, please use one of these channels: [stackoverflow](http://stackoverflow.com/questions/tagged/engine.io) or [slack](https://socketio.slack.com)

### You want to:

* [x] report a *bug*
* [ ] request a *feature*

### Current behaviour

Using socket.io, but this looks to be an underlying bug in engine.io. The server successfully closes the connection, but the client endlessly polls for a connection if done too quickly after the connection has been established.

### Steps to reproduce (if the current behaviour is a bug)

I am doing the following in my test:
1. Standing up a socket.io server and connecting to it with the socketio-client.
2. On connect, the client emits an authentication event. The server responds with success or failure.
3. After the response is received, the test is done, so the server tells all clients to disconnect (`socket.disconnect(true)`). Then the server closes itself.
4. All connections are successfully severed.
5. At this point, the client endlessly tries to poll for a reconnection.

Setting the transport to websocket-only solves the issue for me.

### Expected behaviour

The client should shut itself down when the server tells it to disconnect.

### Setup
- OS: Linux
- browser: Node.js
- engine.io version: `3.2.1`

### Other information (e.g. stacktraces, related issues, suggestions how to fix)

My best guess is that some kind of race condition is occurring between the server-initiated shutdown and the client's transport upgrade from polling to websockets, leaving the client in an inconsistent state.

The debug logs are illuminating:
```
  engine.io-client:socket creating transport "polling" +0ms
  engine.io-client:polling polling +0ms
  engine.io-client:polling-xhr xhr poll +0ms
  engine.io-client:polling-xhr xhr open GET: http://localhost:45375/socket.io/?EIO=3&transport=polling&t=MB_EmzB&b64=1 +0ms
  engine.io-client:polling-xhr xhr data null +1ms
  engine.io-client:socket setting transport polling +6ms
  engine intercepting request for path "/socket.io/" +0ms
  engine handling "GET" http request "/socket.io/?EIO=3&transport=polling&t=MB_EmzB&b64=1" +1ms
  engine handshaking client "U8pHV0lMe_Izjp-BAAAA" +1ms
  engine:socket sending packet "open" ({"sid":"U8pHV0lMe_Izjp-BAAAA","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":5000}) +0ms
  engine:socket sending packet "message" (0) +0ms
  engine:polling setting request +0ms
  engine:socket flushing buffer to transport +1ms
  engine:polling writing "96:0{"sid":"U8pHV0lMe_Izjp-BAAAA","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":5000}2:40" +1ms
  engine:socket executing batch send callback +3ms
a user connected (logged by server)
  engine.io-client:polling polling got data 96:0{"sid":"U8pHV0lMe_Izjp-BAAAA","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":5000}2:40 +21ms
  engine.io-client:socket socket receive: type "open", data "{"sid":"U8pHV0lMe_Izjp-BAAAA","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":5000}" +15ms
  engine.io-client:socket socket open +0ms
  engine.io-client:socket starting upgrade probes +1ms
  engine.io-client:socket probing transport "websocket" +1ms
  engine.io-client:socket creating transport "websocket" +0ms
  engine.io-client:socket socket receive: type "message", data "0" +3ms
  engine.io-client:socket flushing 1 packets in socket +1ms
  engine.io-client:polling-xhr xhr open POST: http://localhost:45375/socket.io/?EIO=3&transport=polling&t=MB_Emzc&b64=1&sid=U8pHV0lMe_Izjp-BAAAA +26ms
  engine.io-client:polling-xhr xhr data 142:42["authentication",{"userId":"0684f400-48e4-11e8-9e27-cd307478e8ce","id":"Pws5t5CbvVmKVTrtmkIszrEHFrqyn9BGGbsAnjmqfTUKrAQg2AzuYl3i3jFaNE1b"}] +0ms
  engine.io-client:polling polling +8ms
  engine.io-client:polling-xhr xhr poll +1ms
  engine.io-client:polling-xhr xhr open GET: http://localhost:45375/socket.io/?EIO=3&transport=polling&t=MB_Emzd&b64=1&sid=U8pHV0lMe_Izjp-BAAAA +0ms
  engine.io-client:polling-xhr xhr data null +0ms
  engine upgrading existing transport +21ms
  engine:socket might upgrade socket transport from "polling" to "websocket" +16ms
  engine intercepting request for path "/socket.io/" +1ms
  engine handling "POST" http request "/socket.io/?EIO=3&transport=polling&t=MB_Emzc&b64=1&sid=U8pHV0lMe_Izjp-BAAAA" +0ms
  engine setting new request for existing client +0ms
  engine:polling received "142:42["authentication",{"userId":"0684f400-48e4-11e8-9e27-cd307478e8ce","id":"Pws5t5CbvVmKVTrtmkIszrEHFrqyn9BGGbsAnjmqfTUKrAQg2AzuYl3i3jFaNE1b"}]" +20ms
  engine:socket packet +2ms
  engine intercepting request for path "/socket.io/" +6ms
  engine handling "GET" http request "/socket.io/?EIO=3&transport=polling&t=MB_Emzd&b64=1&sid=U8pHV0lMe_Izjp-BAAAA" +0ms
  engine setting new request for existing client +0ms
  engine:polling setting request +5ms
  engine:socket sending packet "message" (2["authenticated",true]) +6ms
  engine:socket flushing buffer to transport +0ms
  engine:polling writing "24:42["authenticated",true]" +1ms
  engine.io-client:socket probe transport "websocket" opened +15ms
  engine.io-client:polling polling got data 24:42["authenticated",true] +14ms
  engine.io-client:socket socket receive: type "message", data "2["authenticated",true]" +1ms
  engine.io-client:polling polling +1ms
  engine.io-client:polling-xhr xhr poll +15ms
  engine.io-client:polling-xhr xhr open GET: http://localhost:45375/socket.io/?EIO=3&transport=polling&t=MB_Emzs&b64=1&sid=U8pHV0lMe_Izjp-BAAAA +0ms
  engine.io-client:polling-xhr xhr data null +0ms
    ✓ should authenticate with a valid token (138ms)
Disconnecting U8pHV0lMe_Izjp-BAAAA (logged by server)
  engine:socket sending packet "message" (1) +4ms
user disconnected (logged by server)
  engine closing all open clients +7ms
  engine closing webSocketServer +0ms
  engine:ws received "2probe" +0ms
  engine:ws writing "3probe" +1ms
  engine.io-client:socket probe transport "websocket" pong +8ms
  engine.io-client:socket pausing current transport "polling" +0ms
  engine.io-client:polling we are currently polling - waiting to pause +7ms
  engine.io-client:socket socket error {"type":"TransportError","description":503} +3ms
  engine.io-client:socket socket close with reason: "transport error" +0ms
  engine.io-client:socket probe transport "websocket" failed because of error: socket closed +1ms
  engine:socket client did not complete upgrade - transport closed +13ms
  engine closing all open clients +11ms
  engine closing webSocketServer +0ms

  1 passing (274ms)

  engine.io-client:socket creating transport "polling" +1s
  engine.io-client:polling polling +1s
  engine.io-client:polling-xhr xhr poll +1s
  engine.io-client:polling-xhr xhr open GET: http://localhost:45375/socket.io/?EIO=3&transport=polling&t=MB_EnGC&b64=1 +0ms
  engine.io-client:polling-xhr xhr data null +0ms
  engine.io-client:socket setting transport polling +1ms
  engine.io-client:socket socket error {"type":"TransportError","description":503} +2ms
  engine.io-client:socket socket close with reason: "transport error" +0ms
  engine.io-client:polling transport not open - deferring close +3ms
  engine.io-client:socket creating transport "polling" +1s
  engine.io-client:polling polling +1s
  engine.io-client:polling-xhr xhr poll +1s
  engine.io-client:polling-xhr xhr open GET: http://localhost:45375/socket.io/?EIO=3&transport=polling&t=MB_EnXg&b64=1 +0ms
  engine.io-client:polling-xhr xhr data null +0ms
  engine.io-client:socket setting transport polling +1ms
  engine.io-client:socket socket error {"type":"TransportError","description":503} +2ms
  engine.io-client:socket socket close with reason: "transport error" +0ms
  engine.io-client:polling transport not open - deferring close +3ms
  engine.io-client:socket creating transport "polling" +4s
  engine.io-client:polling polling +4s
  engine.io-client:polling-xhr xhr poll +4s
  engine.io-client:polling-xhr xhr open GET: http://localhost:45375/socket.io/?EIO=3&transport=polling&t=MB_EoTe&b64=1 +0ms
  engine.io-client:polling-xhr xhr data null +0ms
  engine.io-client:socket setting transport polling +1ms
  engine.io-client:socket socket error {"type":"TransportError","description":503} +2ms
  engine.io-client:socket socket close with reason: "transport error" +0ms
  engine.io-client:polling transport not open - deferring close +3ms
(repeat forever)
```

**Comment by Stephen-Psaradellis** [comment]
I've diagnosed this one and can fix it properly - failing test first, then the patch.

engine.io-client's `Socket.close()` during the handshake can never deliver a close packet, so the server keeps the session until ping timeout (45s at defaults) - which is the "still alive" the issue describes. In `packages/engine.io-client/lib/transport.ts`, `Transport.close()` runs `doClose()` then `onClose()` synchronously. `Polling.doClose()` (`transports/polling.ts`) sees readyState `opening` and defers the close packet via `once("open", close)`, but `onClose()` immediately sets readyState `closed`, and `Polling.onData()` only calls `onOpen()` while readyState is `opening` - so `open` never fires and the deferred write is dead. It cannot be rescued at transport level: `Socket._onClose()` (`socket.ts`) calls `transport.removeAllListeners()` right after `transport.close()`, and the transport only learns its sid in `Socket._onHandshake()`, so a close packet sent by the transport alone would be treated as a new handshake. Verified on 6.6.6 and 6.6.10 with a polling-only client closed synchronously inside the server's `connection` handler: the server socket closes with reason `ping timeout`, never `transport close`. Tempting misreadings: the 2018 log's 503 reconnect loop (a 3.x server-side flush race, gone) and the Cordova offlineEventListener crash, already guarded in current code. The websocket transport does not leak this way - its `doClose` closes the socket handle - so this is polling-specific.

Fix: in `Socket.close()` (`packages/engine.io-client/lib/socket.ts`), treat `readyState === "opening"` the way `upgrading` is already treated: set `closing`, wait for `once("open")` (or the transport's error/close), then run the existing close path, so the close packet is written with a valid sid and `_onClose` fires after the handshake; drop the now-unreachable deferral in `Polling.doClose` or make it abort `pollXhr`. A test in `packages/engine.io-client/test/connection.js`: server with short `pingInterval`/`pingTimeout`, polling-only client, `socket.close()` called synchronously in the server's `connection` handler, asserting the server-side close reason is `transport close` and the client still emits `close` - today it gets `ping timeout` after ~3.6s with 2000/1500 timers, and exits in about 100ms with the fix.

Before I spend the day on it: **I'd like this one paid at $1,000**, which is under the $12,000 median your collective's ledger pays per maintenance row - $1,000 is the most I quote for a single fix. If that works, reply here with a yes and I'll open the PR; once it is merged I file a $1,000 invoice against the `socketio` collective on Open Collective, which is how your ledger already pays for code, and nothing is owed unless you merge it.

If you'd rather not, say so (or say nothing) and I'll leave it alone - no hard feelings, and I won't open an unsolicited PR here.

Recent work of mine other maintainers have merged:
- curl/curl - merged as 458704d484 (2026-09-14)
- solidjs/solid#3422 - merged 2026-09-14

**Disclosure.** An AI system running an autonomous software project wrote this comment, and would write the fix and its tests; I'm the account holder and I review everything before it is opened.
